pkg/api/proxy.go

```diff
@@ -0,0 +1,310 @@
+package api
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/dchest/siphash"
+	"github.com/gorilla/mux"
+
+	"zotregistry.dev/zot/pkg/api/config"
+	"zotregistry.dev/zot/pkg/api/constants"
+	"zotregistry.dev/zot/pkg/common"
+)
+
+var ErrFailedToParseIPAddress = errors.New("failed to parse IP address")
+
+// ClusterProxy wraps an http.HandlerFunc which requires proxying between zot instances to ensure
+// that a given repository only has a single writer and reader for dist-spec operations in a scale-out cluster.
+// based on the hash value of the repository name, the request will either be handled locally
+// or proxied to another zot member in the cluster to get the data before sending a response to the client.
+func ClusterProxy(ctrlr *Controller) func(http.HandlerFunc) http.HandlerFunc {
+	return func(next http.HandlerFunc) http.HandlerFunc {
+		return http.HandlerFunc(func(response http.ResponseWriter, request *http.Request) {
+			config := ctrlr.Config
+			logger := ctrlr.Log
+
+			// if no cluster or single-node cluster, handle locally.
+			if config.Cluster == nil || len(config.Cluster.Members) == 1 {
+				next.ServeHTTP(response, request)
+
+				return
+			}
+
+			// since the handler has been wrapped, it should be possible to get the name
+			// of the repository from the mux.
+			vars := mux.Vars(request)
+			name, ok := vars["name"]
+
+			if !ok || name == "" {
+				response.WriteHeader(http.StatusNotFound)
+
+				return
+			}
+
+			// the target member is the only one which should do read/write for the dist-spec APIs
+			// for the given repository.
+			targetMember := computeTargetMember(config, name)
+
+			// from the member list and our DNS/IP address, calculate if this request should be handled locally.
+			// handle locally, if the "address:port" is a match, else, proxy it to the target member.
+			localIPs, err := getLocalServerSockets(config.HTTP.Port)
+			if err != nil {
+				logger.Error().Str(constants.RepositoryLogKey, name).
+					Msg(fmt.Sprintf("failed to get local sockets. error: %s", err.Error()))
+				response.WriteHeader(http.StatusInternalServerError)
+
+				return
+			}
+
+			logger.Debug().Str(constants.RepositoryLogKey, name).Msg(fmt.Sprintf("local sockets: %v", localIPs))
+
+			targetMemberIPs, err := getTargetMemberServerSockets(targetMember)
+			if err != nil {
+				logger.Error().Str(constants.RepositoryLogKey, name).
+					Msg(fmt.Sprintf("failed to get target member sockets. error: %s", err.Error()))
+				response.WriteHeader(http.StatusInternalServerError)
+
+				return
+			}
+
+			logger.Debug().Str(constants.RepositoryLogKey, name).Msg(fmt.Sprintf("target member sockets: %v", targetMemberIPs))
```

Comment:
Avoid %v in Msg()
Instead, use .Str("members", targetMemberIPs).Msg("cluster members")

Same approach everywhere pls.

Reply:
Addressed this. I've changed it to add all the context to the logger as a field as much as possible.